Add temporaryCacheLocation to Cache Options

nicolaszawada opened the pull request:
This is a PR follow up on [the issue we opened a few years ago](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/4231).

So in summary, we added a new property to the `CacheOptions` to make the temporary cache storage location configurable. It defaults to the existing cache storage location so it doesn't break anything when not configured.

nicolaszawada committed:
Add temporaryCacheLocation to Cache Options

nicolaszawada committed:
Change files

nicolaszawada commented:
@microsoft-github-policy-service agree

nicolaszawada committed:
Set correct parameter

nicolaszawada committed:
Merge branch 'dev' into feature/temp-cache-location

tnorling commented on lib/msal-browser/src/cache/BrowserCacheManager.ts:
```diff
@@ -71,16 +70,30 @@ export class BrowserCacheManager extends CacheManager {
                 try {
-                    // Temporary cache items will always be stored in session storage to mitigate problems caused by multiple tabs
-                    return new BrowserStorage(BrowserCacheLocation.SessionStorage);
+                    return new BrowserStorage(tempCacheLocation);
```
I think all of this can be simplified with just the following change: 

```suggestion
                    return new BrowserStorage(temporaryCacheLocation || BrowserCacheLocation.SessionStorage);
```

nicolaszawada replied:
Agreed, the first implementation was more verbose

tnorling commented on lib/msal-browser/src/config/Configuration.ts:
```diff
@@ -77,6 +77,10 @@ export type CacheOptions = {
      */
     cacheLocation?: BrowserCacheLocation | string;
+    /**
+     * Used to specify the temporaryCacheLocation user wants to set. Valid values are "localStorage" and "sessionStorage"
```
memoryStorage?

nicolaszawada replied:
Added

tnorling commented on lib/msal-browser/docs/caching.md:
```diff
@@ -58,7 +58,7 @@ To faciliate efficient token acquisition while maintaining a good UX, MSAL cache
     - performance data
 
-> :bulb: Temporary cache entries will always be stored in session storage or in memory. MSAL will fallback to memory storage if local/session storage is not available.
+> :bulb: Temporary cache entries will always be stored in session storage or in memory unless overridden by the user with the `temporaryCacheLocation` cache option. MSAL will fallback to memory storage if local/session storage is not available.
```
I think we will also want to clearly document the potential problems with overriding the temporary cache location. Specifically when choosing localStorage interaction in more than one tab/window will not be supported and you may receive interaction_in_progress errors unexpectedly. 

Please describe the scenario where this should be overridden. This is an escape hatch, not a fully supported feature.

nicolaszawada replied:
Agreed. I documented this more explicitly now.

tnorling commented on change/@azure-msal-browser-25232dc8-c6d0-4238-ad59-1c1f3fcd1a7a.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "minor",
+  "comment": "Add temporaryCacheLocation to Cache Options",
```
```suggestion
  "comment": "Add temporaryCacheLocation to Cache Options #5725",
```

nicolaszawada replied:
Added

tnorling reviewed (commented):
Thanks for the PR! Left a couple suggestions, would be great if you could include a couple tests as well.

nicolaszawada committed:
Added PR suggestions

nicolaszawada committed:
Add tests

nicolaszawada committed:
Merge branch 'feature/temp-cache-location' of https://github.com/nicolaszawada/microsoft-authentication-library-for-js into feature/temp-cache-location

nicolaszawada committed:
Merge branch 'dev' into feature/temp-cache-location

nicolaszawada commented:
> Thanks for the PR! Left a couple suggestions, would be great if you could include a couple tests as well.

I implemented your suggestions and added some tests.

nicolaszawada committed:
More emphasis in docs

tnorling commented on lib/msal-browser/docs/caching.md:
```diff
@@ -58,10 +58,17 @@ To faciliate efficient token acquisition while maintaining a good UX, MSAL cache
+## Warning :warning:
+Overriding `temporaryCacheLocation` should be done with caution. Specifically when choosing `localStorage`. Interaction in more than one tab/window will not be supported and you may receive `interaction_in_progress` errors unexpectedly. This is an escape hatch, not a fully supported feature.
+
+When using MSAL in a scenario where the user is redirected after successful authentication in a new window or tab, the OAuth 2.0 Authorization Code with PKCE flow will be interrupted. In this case, the original window or tab where the authentication state (code verifier and challenge) are stored, will be lost, and the authentication flow will fail.
```
```suggestion
When using MSAL.js with the default configuration in a scenario where the user is redirected after successful authentication in a new window or tab, the OAuth 2.0 Authorization Code with PKCE flow will be interrupted. In this case, the original window or tab where the authentication state (code verifier and challenge) are stored, will be lost, and the authentication flow will fail.
```

nicolaszawada replied:
Agreed

tnorling commented on lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
+        });
+
+        it(
+            "getTemporaryCache falls back to local storage if not found in session/memory storage",
```
```suggestion
            "getTemporaryCache returns value from localStorage",
```

nicolaszawada replied:
Agreed

tnorling commented on lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
+            expect(browserLocalStorage.getKeys()).toHaveLength(0);
+        });
+
+        describe("Getters and Setters", () => {
```
Tests below this line don't appear relevant to the changes in this PR. Am I missing something?

nicolaszawada replied:
I have copy-pasted these tests from the existing "Interface functions" tests. But in this test setup I changed the cacheConfig for the temporaryCacheLocation to LocalStorage. The reasoning behind this was to show that the existing functionality doesn't break with the temporaryCacheLocation to LocalStorage. If you prefer having them removed. Let me know

tnorling commented on lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
+            }
+        )
+
+        it("setItem", () => {
```
This test doesn't appear relevant to this PR (probably already exists)

nicolaszawada replied:
Agreed

tnorling commented on lib/msal-browser/test/cache/BrowserCacheManager.spec.ts:
```diff
@@ -325,7 +326,532 @@ describe("BrowserCacheManager tests", () => {
+    });
+
+    
+    describe("Interface functions with overridden temporaryCacheLocation", () => {
```
Test additions above this line don't appear relevant to this PR

nicolaszawada replied:
I have copy-pasted these tests from "Interface functions" tests. But in this test setup I changed the cacheConfig for the temporaryCacheLocation to LocalStorage. The reasoning behind this was to show that the existing functionality doesn't break with the temporaryCacheLocation to LocalStorage.

nicolaszawada replied:
If you prefer having them removed, let me know

nicolaszawada replied:
These tests were already there

nicolaszawada replied:
These tests were already there. I just moved them in this file in a more logical order

tnorling reviewed (commented):
Thanks for addressing the feedback, code and docs look good. I'm a little confused about some of the tests, hoping you can clear that up. Thanks!

nicolaszawada committed:
Implemented tnorling feedback

nicolaszawada commented:
> Thanks for addressing the feedback, code and docs look good. I'm a little confused about some of the tests, hoping you can clear that up. Thanks!

Thanks for your time & feedback, I have implemented your suggestions and cleared some things up. Let me know if there's anything else I can do.

nicolaszawada commented:
@tnorling, I hope you're doing well. I understand that you may be busy with other tasks, but I just wanted to remind you that the PR is still open and awaiting review. If there's anything I can do to help, please don't hesitate to let me know.
Thank you for your time and consideration. I look forward to hearing back from you soon.

tnorling reviewed (approved):
Thanks for the reminder and addressing my feedback!

nicolaszawada commented:
Thanks for the approval!
Hm I notice the msal-browser CI failing due to a timeout. I don't really see the issue. Is there anything I can do?

tnorling committed:
Merge branch 'dev' into feature/temp-cache-location

tnorling commented:
> 

Don't worry about it, tests have been a bit flakey lately, we'll sort it out

nicolaszawada commented:
> > 
> 
> Don't worry about it, tests have been a bit flakey lately, we'll sort it out

Hi @tnorling, is there any update on the above please?

tnorling committed:
Merge branch 'dev' into feature/temp-cache-location

sameerag committed:
Merge branch 'dev' into feature/temp-cache-location

jo-arroyo committed:
Update package-locks

jo-arroyo committed:
Merge branch 'dev' of https://github.com/AzureAD/microsoft-authentication-library-for-js into dev

hectormmg committed:
MSAL Major Version Upgrades April 2023 (#5771)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Shane Oatman <shoatman@microsoft.com>
Co-authored-by: Thomas Norling <thomas.l.norling@gmail.com>
Co-authored-by: konstantin-msft <119362382+konstantin-msft@users.noreply.github.com>
Co-authored-by: Konstantin Shabelko <kshabelko@microsoft.com>
Co-authored-by: Jo Arroyo <joarroyo@microsoft.com>
Co-authored-by: Robbie-Microsoft <87724641+Robbie-Microsoft@users.noreply.github.com>
Co-authored-by: Robert Ginsburg <rgins16@gmail.com>
Co-authored-by: Thomas Norling <thnorlin@microsoft.com>
Co-authored-by: Doğan Erişen <v-derisen@microsoft.com>
Co-authored-by: Lalima Sharda <lalima.sharda@gmail.com>
Co-authored-by: Sameera Gajjarapu <sameera.gajjarapu@microsoft.com>
Co-authored-by: Lalima Sharda <lalimasharda@microsoft.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: konstantin-msft <konstantin-msft@users.noreply.github.com>
Co-authored-by: Salman Salem <v-ssalem@microsoft.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: tnorling <tnorling@users.noreply.github.com>

dependabot[bot] committed:
Bump peter-evans/create-pull-request from 4 to 5 (#5866)

Bumps
[peter-evans/create-pull-request](https://github.com/peter-evans/create-pull-request)
from 4 to 5.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/peter-evans/create-pull-request/releases">peter-evans/create-pull-request's
releases</a>.</em></p>
<blockquote>
<h2>Create Pull Request v5.0.0</h2>
<h2>Behaviour changes</h2>
<ul>
<li>The action will no longer leave the local repository checked out on
the pull request <code>branch</code>. Instead, it will leave the
repository checked out on the branch or commit that it was when the
action started.</li>
<li>When using <code>add-paths</code>, uncommitted changes will no
longer be destroyed. They will be stashed and restored at the end of the
action run.</li>
</ul>
<h2>What's new</h2>
<ul>
<li>Adds input <code>body-path</code>, the path to a file containing the
pull request body.</li>
<li>At the end of the action run the local repository is now checked out
on the branch or commit that it was when the action started.</li>
<li>Any uncommitted tracked or untracked changes are now stashed and
restored at the end of the action run. Currently, this can only occur
when using the <code>add-paths</code> input, which allows for changes to
not be committed. Previously, any uncommitted changes would be
destroyed.</li>
<li>The proxy implementation has been revised but is not expected to
have any change in behaviour. It continues to support the standard
environment variables <code>http_proxy</code>, <code>https_proxy</code>
and <code>no_proxy</code>.</li>
<li>Now sets the git <code>safe.directory</code> configuration for the
local repository path. The configuration is removed when the action
completes. Fixes issue <a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1170">peter-evans/create-pull-request#1170</a>.</li>
<li>Now determines the git directory path using the <code>git rev-parse
--git-dir</code> command. This allows users with custom repository
configurations to use the action.</li>
<li>Improved handling of the <code>team-reviewers</code> input and
associated errors.</li>
</ul>
<h2>News</h2>
<p>:trophy: create-pull-request won <a
href="https://twitter.com/peterevans0/status/1638463617686470657?s=20">an
award</a> for &quot;awesome action&quot; at the Open Source Awards at
GitHub Universe. Thank you for your support and for making
create-pull-request one of the top used actions. Please give it a
:star:, or even <a href="https://github.com/sponsors/peter-evans">buy me
a coffee</a>.</p>
<h2>What's Changed</h2>
<ul>
<li>v5 by <a
href="https://github.com/peter-evans"><code>@​peter-evans</code></a> in
<a
href="https://redirect.github.com/peter-evans/create-pull-request/pull/1792">peter-evans/create-pull-request#1792</a></li>
<li>15 dependency updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/peter-evans/create-pull-request/compare/v4.2.4...v5.0.0">https://github.com/peter-evans/create-pull-request/compare/v4.2.4...v5.0.0</a></p>
<h2>Create Pull Request v4.2.4</h2>
<p>⚙️ Patches some recent security vulnerabilities.</p>
<h2>What's Changed</h2>
<ul>
<li>Update concepts-guidelines.md by <a
href="https://github.com/chrisbruford"><code>@​chrisbruford</code></a>
in <a
href="https://redirect.github.com/peter-evans/create-pull-request/pull/1610">peter-evans/create-pull-request#1610</a></li>
<li>58 dependency updates by <a
href="https://github.com/dependabot"><code>@​dependabot</code></a></li>
</ul>
<h2>New Contributors</h2>
<ul>
<li><a
href="https://github.com/chrisbruford"><code>@​chrisbruford</code></a>
made their first contribution in <a
href="https://redirect.github.com/peter-evans/create-pull-request/pull/1610">peter-evans/create-pull-request#1610</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/peter-evans/create-pull-request/compare/v4.2.3...v4.2.4">https://github.com/peter-evans/create-pull-request/compare/v4.2.3...v4.2.4</a></p>
<h2>Create Pull Request v4.2.3</h2>
<h2>What's Changed</h2>
<ul>
<li>fix: add check for missing token input by <a
href="https://github.com/peter-evans"><code>@​peter-evans</code></a> in
<a
href="https://redirect.github.com/peter-evans/create-pull-request/pull/1324">peter-evans/create-pull-request#1324</a></li>
</ul>
<p><strong>Full Changelog</strong>: <a
href="https://github.com/peter-evans/create-pull-request/compare/v4.2.2...v4.2.3">https://github.com/peter-evans/create-pull-request/compare/v4.2.2...v4.2.3</a></p>
<h2>Create Pull Request v4.2.2</h2>
<h2>What's Changed</h2>
<ul>
<li>fix: support github server url for pushing to fork by <a
href="https://github.com/peter-evans"><code>@​peter-evans</code></a> in
<a
href="https://redirect.github.com/peter-evans/create-pull-request/pull/1318">peter-evans/create-pull-request#1318</a></li>
</ul>
<!-- raw HTML omitted -->
</blockquote>
<p>... (truncated)</p>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/5b4a9f6a9e2af26e5f02351490b90d01eb8ec1e5"><code>5b4a9f6</code></a>
v5 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1792">#1792</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/1847e5d1d66115e99988ff4ccbe629ae5b14820b"><code>1847e5d</code></a>
build(deps-dev): bump eslint from 8.36.0 to 8.37.0 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1803">#1803</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/c246f7e9126aba0f1cc780a7133429d9771e9b66"><code>c246f7e</code></a>
build(deps-dev): bump <code>@​typescript-eslint/parser</code> from
5.57.0 to 5.57.1 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1801">#1801</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/2dd2b11b09f0b3bf86c25cf615e4bbe624f80314"><code>2dd2b11</code></a>
build(deps-dev): bump eslint-import-resolver-typescript (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1802">#1802</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/05d5a3c3f9506f8628125834beec3993d3eada67"><code>05d5a3c</code></a>
build(deps-dev): bump <code>@​types/node</code> from 18.15.10 to
18.15.11 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1800">#1800</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/21479f22fcc890fd56c1b5a47d47162fa907dddc"><code>21479f2</code></a>
build(deps-dev): bump ts-jest from 29.0.5 to 29.1.0 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1799">#1799</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/36a56dac0739df8d3d8ebb9e6e41026ba248ec27"><code>36a56da</code></a>
build(deps-dev): bump <code>@​typescript-eslint/parser</code> from
5.56.0 to 5.57.0 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1768">#1768</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/b7f0c9773b3f9bd0eb8d00cc963aceb9da2345e5"><code>b7f0c97</code></a>
build(deps-dev): bump prettier from 2.8.6 to 2.8.7 (<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1767">#1767</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/6a62596740af10ccfc9800db9f74c255f65c9055"><code>6a62596</code></a>
build(deps): bump peter-evans/enable-pull-request-automerge from 2 to 3
(<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1766">#1766</a>)</li>
<li><a
href="https://github.com/peter-evans/create-pull-request/commit/d1ed29fe1ea9695e51f6bdda3db7211f3294d9d5"><code>d1ed29f</code></a>
build(deps-dev): bump <code>@​types/node</code> from 18.15.5 to 18.15.10
(<a
href="https://redirect.github.com/peter-evans/create-pull-request/issues/1765">#1765</a>)</li>
<li>Additional commits viewable in <a
href="https://github.com/peter-evans/create-pull-request/compare/v4...v5">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=peter-evans/create-pull-request&package-manager=github_actions&previous-version=4&new-version=5)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tnorling committed:
Merge branch 'msal-lts' into feature/temp-cache-location

tnorling committed:
passport-azure-ad update node-jose dependency (#5869)

Updates node-jose dependency to 2.2.0+ to address vulnerability with
versions < 2.2.0

tnorling committed:
Update PoliCheck version (#5864)

Update PoliCheck version as the one we're using now is being deprecated

konstantin-msft committed:
Fix telemetry typos (#5868)

- Remove `startPerformanceMeasuremeant` that is a duplicate of
`startPerformanceMeasurement`.
- Rename `astsAsyncMeasurement` -> `atsAsyncMeasurement`.

nicolaszawada committed:
Merge branch 'dev' of https://github.com/nicolaszawada/microsoft-authentication-library-for-js into feature/temp-cache-location

nicolaszawada committed:
Merge branch 'feature/temp-cache-location' of https://github.com/nicolaszawada/microsoft-authentication-library-for-js into feature/temp-cache-location

nicolaszawada committed:
Fix BrowserCacheManager tests

nicolaszawada committed:
Revert "Fix telemetry typos (#5868)"

This reverts commit 9f5c1288cbe52fa2176f41269032c7bb7920de14.

nicolaszawada committed:
Revert "Update PoliCheck version (#5864)"

This reverts commit 629883fbfe597b56260483e4567bacd9ae9ba81d.

nicolaszawada committed:
Revert "passport-azure-ad update node-jose dependency (#5869)"

This reverts commit 11cdfc901fd35bbb097295d41eba31bfd1447588.

nicolaszawada committed:
Revert "Bump peter-evans/create-pull-request from 4 to 5 (#5866)"

This reverts commit d929cd7d7e889cea3899c6d8603efaa38e7d57ca.

nicolaszawada committed:
Revert "MSAL Major Version Upgrades April 2023 (#5771)"

This reverts commit c0c956a1d31ca5fd4218e04616911a96eac37ba6.

nicolaszawada committed:
Revert "Update package-locks"

This reverts commit da5f53543cd97f6d1e0c180032d4f7d010faafa6.

nicolaszawada committed:
Revert "Post-release April 2023 (#5850)"

This reverts commit c4d8d227b18e7c6585502ca17e687f09db99cfb8.

nicolaszawada committed:
Revert "Revert "Update package-locks""

This reverts commit 5f4f46100667444378092423fb1e15640c870b39.

nicolaszawada committed:
Fix MD files

nicolaszawada commented:
Hi @tnorling, I saw that there were some tests failing due to the fact that they were added **after** this PR and didn't include the new `temporaryCacheLocation` property. I have fixed the failing tests.

I noticed too late that you changed the target branch from `dev` to `msal-lts`, so I had to do some reverting cleanup of some `dev` commits that would have been merged into `msal-lts`. I think it would be better to squash commit this PR to not clutter the history.

tnorling committed:
Merge branch 'msal-lts' into feature/temp-cache-location

tnorling commented:
@nicolaszawada Thanks for fixing the tests and my apologies for the delay in merging

tnorling committed:
Merge branch 'msal-lts' into feature/temp-cache-location

ghost commented:
:tada:`@azure/msal-browser@v2.37.0` has been released which incorporates this pull request.:tada:

Handy links:
* [Release Notes](https://github.com/AzureAD/microsoft-authentication-library-for-js/releases/tag/msal-browser-v2.37.0)
* [Official NPM Release](https://www.npmjs.com/package/@azure/msal-browser/v/2.37.0)

ghost commented:
:tada:`@azure/msal-browser@v3.0.0-alpha.0` has been released which incorporates this pull request.:tada:

Handy links:
* [Release Notes](https://github.com/AzureAD/microsoft-authentication-library-for-js/releases/tag/msal-browser-v3.0.0-alpha.0)
* [Official NPM Release](https://www.npmjs.com/package/@azure/msal-browser/v/3.0.0)
